### ISSUE TITLE
USB-OTG support (ttyACM0) on Kobo mini not available since Version 7

### ISSUE THREAD

**Issue by BiGolden** [XCSoar/XCSoar#879]
XCSoar versions having and not having the problem
-------------------------------------------------

 
Problem exists in 7.23, 7.16, 7.5 (tested in thie range) on Kobo mini
No Problem for versions 6.8.xx (last tested 6.8.17) on Kobo mini 
(was ok even on Kobo Glo, althougt OTG on this device is not certified by XCSoar)

Expected behavior
-----------------

Reading serial data (Vario, Flarm) from different devices (XTracer mini II, TTGO T-Bar with softRF) via Kobo USB-OTG was working perfectly after switching "Enable USB-OTG" under "System" in start menue, rebooting and selecting port ttyACM0 in the device list. 

Actual behavior
---------------

The port ttyACM0 is not offered any more in the device list in Version 7.xx (used starting procedure exactly the same as in 6.8.17). 

Steps to reproduce the behavior
-------------------------------

Tested for 6.8.16, 6.8.17 (functioning) and 7.5, 7.16, 7.23 (not functioning)

Do you have any idea what may have caused this?
-----------------------------------------------

Switching from "show all existing devices" to "show only available devices" might have caused the problem.
(see https://github.com/XCSoar/XCSoar/commit/2de78d4aa22cee2913db0031befa0f51019e3d2b)

Do you have an idea how to solve the issue?
-------------------------------------------

I feel, that the procedure how available devices are searched for on the Kobo (udevd?) might miss the OTG-device switched on before under System......

**Comment by cmoetzing** [comment]
Same with Stodeus leGPSBip+. 6.8.17 is working and 7.23 is not. I even have the same device entry in the NMEA list but can't connect since it says ttycm0 is not available.

**Comment by tookayNET** [comment]
I've also noticed that when in the XCSoar Home Screen on my Kobo, and going in to the "System Menu", the "Enable/Disable USB-OTG" does toggle from "Enable" and "Disable" after pressing. If I reboot after pressing it and come back to this menu I will see the change reflected, but it doesn't happen "live" so to speak.  Could this be related, in that XC Soar no longer correctly "sees" the USB-OTG status so does not display ttyACM0?

**Comment by GeekGoneOld** [comment]
> 've also noticed that when in the XCSoar Home Screen on my Kobo, and going into the "System Menu", the "Enable/Disable USB-OTG" doesn't toggle from "Enable" and "Disable" after pressing.

I believe all Kobos EXCEPT Clara HD must reboot to switch USB-OTG on or off.  Only Clara HD can reboot on the fly.  Problem seems to be the lack of reboot which is supposed to happen automatically after you click on the {Enable | Disable} USB-OTG button.

**Comment by BiGolden** [comment]
> I believe all Kobos EXCEPT Clara HD must reboot to switch USB-OTG on or off. Only Clara HD can reboot on the fly. Problem seems to be the lack of reboot which is supposed to happen automatically after you click on the {Enable | Disable} USB-OTG button.

I agree that Kobo Mini and Glo won't reboot on the fly. You have to reboot manually. However the problem of non accessible ttyACM0 exists even if you manually reboot after switching OTG on. So this behaviour ist not the reaseon for the problem.

**Comment by BiGolden** [comment]
> I believe all Kobos EXCEPT Clara HD must reboot to switch USB-OTG on or off. Only Clara HD can reboot on the fly. Problem seems to be the lack of reboot which is supposed to happen automatically after you click on the {Enable | Disable} USB-OTG button.

I agree that Kobo Mini and Glo won't reboot on the fly. You have to reboot manually. However the problem of non accessible ttyACM0 exists even if you manually reboot after switching OTG on. So this behaviour ist not the reaseon for the problem.

**Comment by GeekGoneOld** [comment]
If this is same as #888 (i.e. port exists but doesn't show up in list), a workaround is to define your device using any tty port then close XCSoar, edit your default.prf (or whatever .prf you use) and change the port name from what you chose to ttyACM0

**Comment by GeekGoneOld** [comment]
Please verify using telnet if the port exists.  If it does, I believe this is identical to #888.  If so, I noticed that if I changed src/Device/Port/TTYEnumerator.cpp CheckTTYName to filter 'n" to 'z" instead of 'p' to 'w' it works.  I have NOT submitted a PR because this isn't really getting to the root of the problem and affects other target platforms.
`@@ -40,7 +40,7 @@ CheckTTYName(const char *name) noexcept

     /* ignore virtual internal ports on Mac OS X (and probably other
        BSDs) */
-    if (*t >= 'p' && *t <= 'w')
+    if (*t >= 'n' && *t <= 'z')
       return false;
`

**Comment by cmoetzing** [comment]
I would say (without testing) that the device ttyACM0 does not exist (in my case). My vario makes a noise when USB OTG connects and on 7 it does not do that. Also I still had the port entry in NMEA list (from 6.8) and I could not connect it. Will try to verify on the device the next days.

**Comment by GeekGoneOld** [comment]
> I would say (without testing) that the device ttyACM0 does not exist (in my case). My vario makes a noise when USB OTG connects and on 7 it does not do that. Also I still had the port entry in NMEA list (from 6.8) and I could not connect it. Will try to verify on the device the next days.

When you test, could you look with telnet to see if the file /dev/ttyACM0 is present on the Kobo?  Could you count how many ports are in the list?

**Comment by cmoetzing** [comment]
`/dev/ttyACM0` is not present.

```
~ # ls /dev/tty*
/dev/tty      /dev/tty13    /dev/tty19    /dev/tty24    /dev/tty3     /dev/tty35    /dev/tty40    /dev/tty46    /dev/tty51    /dev/tty57    /dev/tty62    /dev/ttymxc1
/dev/tty0     /dev/tty14    /dev/tty2     /dev/tty25    /dev/tty30    /dev/tty36    /dev/tty41    /dev/tty47    /dev/tty52    /dev/tty58    /dev/tty63    /dev/ttymxc2
/dev/tty1     /dev/tty15    /dev/tty20    /dev/tty26    /dev/tty31    /dev/tty37    /dev/tty42    /dev/tty48    /dev/tty53    /dev/tty59    /dev/tty7     /dev/ttyp0
/dev/tty10    /dev/tty16    /dev/tty21    /dev/tty27    /dev/tty32    /dev/tty38    /dev/tty43    /dev/tty49    /dev/tty54    /dev/tty6     /dev/tty8     /dev/ttyp1
/dev/tty11    /dev/tty17    /dev/tty22    /dev/tty28    /dev/tty33    /dev/tty39    /dev/tty44    /dev/tty5     /dev/tty55    /dev/tty60    /dev/tty9     /dev/ttyp2
/dev/tty12    /dev/tty18    /dev/tty23    /dev/tty29    /dev/tty34    /dev/tty4     /dev/tty45    /dev/tty50    /dev/tty56    /dev/tty61    /dev/ttymxc0  /dev/ttyp3
```
I am not sure which list so here we go:

1. Slots in NMEA Dialog "A-F"
2. Dialog for one NMEA slot -> Port: 7 (Deactivated, TCP-Client, TCP-Port, UDP-Port, ttymxc0, ttymxc1, ttymxc2)

**Comment by lordfolken** [comment]
Can you unplug and replug your device and paste the most recent kernel messages?  dmesg is the Command.

**Comment by cmoetzing** [comment]
I re-did everything (flash 7.23, enable OTG, reboot, plug-in device) and now it works. `/dev/ttyACM0` present and device is recognized.
Not sure where I went wrong the last times (pretty sure I did the same thing several times, especially since I confirmed it working with 6.8). Sorry for the confusion. I noticed the reboot button is not working though. Had to manually power off and on.
@BiGolden maybe re-flash aswell and enable OTG with manaul reboot and try again?

**Comment by BiGolden** [comment]
> @BiGolden maybe re-flash aswell and enable OTG with maual reboot and try again?

Meanwhile I tried almost every configuration: Kobo Mini, Kobo Glo, reflashing and rebooting manually multiple times but unfortunately no success. I thought it could be that Kobo does not recognize OTG while rebooting when no device is connected since Kobo does not provide any voltage on USB. So I redid everything with XTracer connected (USB-Voltage supported) - no succes. Needless to say that the device is to be seen in the list when you come from 6.8 or manually edit the *.prf-file changing or adding the port name to /dev/ttyACM0. However the port still is not accessible (message "failed to open /dev/ttyACM0: No such file or device") and if you switch to any other device the port dissappears from the list. I'm sorry to have no more idea.....

**Comment by tookayNET** [comment]
Likewise.

I've re-flashed 7.23 on my Kobo Aura with no luck.
Telneting into the /dev directory does not show ttyACM0, though there is a tty0 through to tty9, ttymxc0, and ttyp0 to ttyp3.

**Comment by GeekGoneOld** [comment]
First, this is not the same as #888 as in this one the port doesn’t exist and in the other it exists but doesn’t get put in the selection list

Second, based on above, this appears to fail with mini, Glo and Aura 2nd Ed.  It likely works with Clara HD as I know I’ve used ttyACM0 on 7.23+driver (release was missing Driver’s).  Note: @cmoetzing  did not disclose his platform but it failed only intermittently.

**Comment by BiGolden** [comment]
> 4\. The target platforms are failing to reboot on the switch to/from OTG. This problem should be solved first!!!

Many thanks to GeekGoneOld for the interim balance!
However, on my recent tests I found, that Kobo mini does in fact reboot automatically on the switch to/from OTG - I was wrong with my comment above ('I agree that Kobo Mini and Glo won't reboot on the fly ...'). I'm sorry about that! 
It's only the Glo (1st edition) which does not, however Glo is not officially supported as I mentioned at the beginning of the thread. There is a message on the Glo
"This feature was designed for the Kobo Mini, Touch 2.0, Glo HD and Aura 2, but this is not one. Use at your own risk.".

Therefore point 4. could be skipped on my opinion and I'd propose to concentrate on the mini only to avoid any confusion.

**Comment by cmoetzing** [comment]
@GeekGoneOld Kobo Glo for me.

**Comment by cmoetzing** [comment]
Just wondering: file `SystemDialog.cpp` checks Kobo Glo but does not list it in the message. Is this on purpose? Also Clara HD is missing.

```
  if (model != KoboModel::MINI &&
      model != KoboModel::GLO &&
      model != KoboModel::TOUCH2 &&
      model != KoboModel::GLO_HD &&
      model != KoboModel::AURA2 &&
      ShowMessageBox(_T("This feature was designed for the Kobo Mini, Touch 2.0, Glo HD and Aura 2, but this is not one.  Use at your own risk.  Continue?"),
                     _T("USB-OTG"), MB_YESNO) != IDYES)
    return;
```

**Comment by GeekGoneOld** [comment]
> @GeekGoneOld Kobo Glo for me.

Thanks.  Just to be very specific, Glo, not Glo HD?  @BiGolden same with you?

> Just wondering: file `SystemDialog.cpp` checks Kobo Glo but does not list it in the message. Is this on purpose? Also Clara HD is missing.

Only 3 kernels were modified for OTG: Aura2, Glo HD and "Kobo".  Clara HD does not switch kernels at all (hence missing from list and text), Touch2 uses the same kernel as Glo HD.  All others use "Kobo" and, presumably, that means that Glo was known to work with "Kobo".  I suspect Glo is missing from the text because it was just found to work using "Kobo", not designed to work.  Sort of a CYA?

In any event, the kernels worked with 6 so they should work with 7.  I think the problem is elsewhere.

> However, on my recent tests I found, that Kobo mini does in fact reboot automatically on the switch to/from OTG

That makes a lot of sense.  Hopefully the others were mistaken as well. @BiGolden , @tookayNET ?

**Comment by tookayNET** [comment]
> > However, on my recent tests I found, that Kobo mini does in fact reboot automatically on the switch to/from OTG
> 
> That makes a lot of sense. Hopefully the others were mistaken as well. @BiGolden , @tookayNET ?

No, my Aura 2 never reboots on switching to/from OTG. 
In fact, even the "reboot" button doesn't work.
I have to choose "Poweroff" at the main XCSoar menu in order for the disable/enable USB-OTG action to be reflected.

**Comment by BiGolden** [comment]
> Thanks. Just to be very specific, Glo, not Glo HD? @BiGolden same with you?
Correct. I own a Glo, NOT Glo HD.

> I have to choose "Poweroff" at the main XCSoar menu in order for the disable/enable USB-OTG action to be reflected.
Same with the Glo.

**Comment by cmoetzing** [comment]
I have a Kobo Glo N613 (non HD).
I can tap reboot as often as I want nothing is happening. 

> Clara HD does not switch kernels at all (hence missing from list and text)

For the end user this distinction is irrelevant I guess? The button is tappable on the Clara and the user gets a warning that is not relevant as the Clara has a dedicated implementation for OTG, right?

**Comment by cmoetzing** [comment]
On the Glo via telnet:
```
~ # /sbin/reboot
~ # echo $?
0
```
Nothing happens.
Kobo Clara does as `/sbin/reboot -f` which works also for the Glo via `telnet`.

**Comment by GeekGoneOld** [comment]
USB has two types of connections: 1. Host (master) which is typically the “computer”; 2. Gadget (slave) typically a thumb drive , keyboard, dongle of any sort (speaker, serial cable etc) and so on. 

A phone or tablet USB port is a gadget by default (I.e. connect phone to computer and phone looks like a thumb drive). 

To allow you to connect a serial cable to a phone (or Kobo) you need the port to be a host port. To do this you need OTG in the kernel AND you have to tell OTG whether the port should be in host or gadget mode.

Every Kobo except Clara HD does NOT have OTG in the kernel so the XCSoar team made kernels with OTG and defaulting to host mode. Since nickel needs the factory kernel, the kernel is swapped to switch modes. 

Since factory Clara HD kernel HAS OTG, modes are switched without touching the kernel. 

In answer to your question, Clara HD never gets to the warning pop up because it never tries to switch kernel. Look at where switch kernel is called from and you will see that Clara HD is specifically excluded.  It does, however, actually do something when you tap on the OTG choice and immediately updates the text as the mode switch is effectively instant.

**Comment by GeekGoneOld** [comment]
Yes. I was strongly suspecting that. Clara HD will not work without -r but the dirt release supporting Clara HD ensured it had -r but didn’t alter other models which were assumed to work  I will submit a PR to use -r for all models. Not sure why reboot stopped working or if it ever did but -r may fix that tiny problem. 

I strongly doubt that will fix the port problem, though.

**Comment by GeekGoneOld** [comment]
As to whether this is fixed by PR1111: Absolute no but also maybe.

The OP (@BiGolden) was using a Kobo Mini which requires a reboot to enable OTG  so I suggested lack of reboot was the problem but @BiGolden replies with a definitive no:

> I agree that Kobo Mini and Glo won't reboot on the fly. You have to reboot manually. However the problem of non accessible ttyACM0 exists even if you manually reboot after switching OTG on. So this behaviour ist not the reaseon for the problem.

He later said he was mistaken and the reboot was happening (on the Mini) automatically.  Therefore, as far as the OP goes, absolutely not connected to PR1111.  It looks like it could be a power/cabling issue between the Kobo and the XTracer.  I'd suggest trying an FTDI serial cable with indication that it is powered.

BUT

There were two other posters (@cmoetzing  and @tookayNET ) who posted different problems which have not been addressed and the OP also tried a Glo HD and it seems it was experiencing a reboot issue which, yes, is solved by PR1111.  Looks like on problem was fixed by a reload from scratch (@cmoetzing) and there may be other problems but they weren't the OP so they really should have their own issue(s) if they persist.

SO

The original post is still unsolved as it is not solved by PR1111 and some of the other issues mentioned that are different from the OP may be unsolved but should be moved to their own issue.
